### PR TITLE
feat(node:timers): polyfill `node:timers` module

### PR DESCRIPTION
**Pull request by jculvey**
Replaces the current auto-mocking of 'timers' to support destructured ESM imports and allow for functional polyfill coverage in the future.

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/runtime/node/timers/promises/index.ts
Comment (IgorMinar) [inline src/runtime/node/timers/promises/index.ts:5]: I think this makes sense since setImmediate is quite problematic for runtimes to implement so it's native availability is spotty.

We could try to shim it, but it's challenging because setImmediate is something in between queueMicrotask and setTimeout: https://nodejs.org/en/learn/asynchronous-work/event-loop-timers-and-nexttick#setimmediate-vs-settimeout

Some have used `postMessage` to get closer to the real behavior (e.g. [vscode here](https://github.com/microsoft/vscode/blob/e045e3b1b25de81232b7db32747f31bf90d02d13/src/vs/base/common/platform.ts#L193-L193)) but even then there are issues, proving this to be a problematic api to shim: https://github.com/microsoft/vscode/issues/133173

I'm ok throwing for now and leaving a comment in the code about these considerations and come back to it later once the need arises.

**Review by IgorMinar — APPROVED** [review]
a few tweaks needed but over LGTM! thank you

File: src/runtime/node/timers/index.ts
Comment (pi0) [inline src/runtime/node/timers/index.ts:20]: I was thinking in js runtimes that timeouts don't make sense, perhaps we could immediately call the function as fallback? (if goal is compatibility, calling is better than throwing IMO)

File: src/runtime/node/timers/promises/index.ts
Comment (pi0) [reply src/runtime/node/timers/promises/index.ts:5]: Maybe we could try a proimisified version of default `setImmediate` and fallback to queueMicrotask? If the end-goal is maximum compatibility, fallback behaviors (even if not precise) IMO would be better than throwing errors

Comment (IgorMinar) [reply src/runtime/node/timers/promises/index.ts:5]: queueMicrotask will result in ordering being different in the polyfilled version. a postMessage implementation is likely the closes we can get to the real thing, unless we actually implement this in `workerd` natively.

File: src/runtime/node/timers/index.ts
Comment (IgorMinar) [reply src/runtime/node/timers/index.ts:20]: I understand the desire for getting the code working, but setTimeout is such a core api that turning it from an async to async call will likely cause more harm than good. I think it's highly unlikely that we'll ever see a runtime without `setInterval` and `setTimeout` in practice. So for these particular APIs we might as well drop the notImplementedAsync fallback to save a few bytes.

Comment (pi0) [reply src/runtime/node/timers/index.ts:20]: Quick search: QuickJS and MuJS don't implement timers by default and such limited situations where runtimes don't have timers honestly make sense to me similar to work that sets `Date.now()` to `0` in ambient contexts.

PS: Let's consider when Node.js targeted libraries are used in build, they often pull in lots of more polyfills neither are bundle optimized, that few bytes is precision errors.

Comment (jculvey) [reply src/runtime/node/timers/index.ts:20]: What's the verdict here? `setTimeout || throw` or `setTimeout || call function as fallback`?

File: src/runtime/node/timers/promises/index.ts
Comment (pi0) [reply src/runtime/node/timers/promises/index.ts:5]: promisified version of setImmediate could be likely used in in libraries to make a delay. Exposing an implementation from native scheduler of workerd runtime would be amazing. In the meantime i suggest to simply promisify the global setImmediate to give basic functionality rather than breaking.

File: src/runtime/node/timers/index.ts
Comment (pi0) [reply src/runtime/node/timers/index.ts:23]: Yep, and I think using second is totally safe for workerd at least..

File: src/runtime/node/timers/promises/index.ts
Comment (IgorMinar) [reply src/runtime/node/timers/promises/index.ts:5]: we could do that if setImmediate is present, but keep in mind that many runtimes including deno and workerd don't support it.

File: src/runtime/node/timers/index.ts
Comment (IgorMinar) [reply src/runtime/node/timers/index.ts:23]: I'm not convinced about calling the fn as a fallback directly, but since won't have any impact on workerd, I'll defer to @pi0's decision here.

File: src/runtime/node/timers/promises/index.ts
Comment (pi0) [reply src/runtime/node/timers/promises/index.ts:5]: Yeah it would be anyway not consistent with Node.js core and i belive runtime-specific Node polyfills would be different...

I think unenv could simply try it best to make things working on best-effort basis at best.

Checking Deno 1.44.0 it seems to work:

image

And bun returning promise with stagenly not calling cb or somehow a bug:

image

Comment (pi0) [reply src/runtime/node/timers/promises/index.ts:5]: Let's defer this discusison

File: src/runtime/node/timers/index.ts
Comment (pi0) [reply src/runtime/node/timers/index.ts:23]: Let's defer this until we need

**Review by pi0 — APPROVED** [review]
Thanks! (defered promises and timerless runtimes discussions until we hit the cases we need)

**Pull request merged**
